src/core.rs

```diff
@@ -326,19 +343,82 @@ impl Fragment for Word<'_> {
+/// regions of space characters (`' '`). This works for Western
+/// languages without emojis and similar non-ASCII characters. A more
+/// complex way of finding words is to use the Unicode line breaking
+/// algorithm described in the [Unicode Standard Annex
+/// #14](https://www.unicode.org/reports/tr14/). This algorithm finds
+/// break points in non-ASCII text.
 ///
 /// # Examples
 ///
+/// Finding words by whitespace:
+///
 /// ```
-/// use textwrap::core::{find_words, Fragment, Word};
-/// let words = find_words("Hello World!").collect::<Vec<_>>();
+/// use textwrap::core::{find_words, Fragment, LineBreakAlgorithm, Word};
+///
+/// let words = find_words("Hello World!", LineBreakAlgorithm::Whitespace);
 /// assert_eq!(words, vec![Word::from("Hello "), Word::from("World!")]);
 /// assert_eq!(words[0].width(), 5);
 /// assert_eq!(words[0].whitespace_width(), 1);
 /// assert_eq!(words[0].penalty_width(), 0);
 /// ```
-pub fn find_words(line: &str) -> impl Iterator<Item = Word> {
+///
+/// Using the Unicode line breaking algorithm, we can find line break
+/// opportunities between characters with no interspaced whitespace:
+///
+/// ```
+/// use textwrap::core::{find_words, LineBreakAlgorithm, Word};
+///
+/// #[cfg(feature = "unicode-linebreak")]
+/// assert_eq!(find_words("Emojis: 😂😍", LineBreakAlgorithm::UnicodeLineBreaks),
+///            vec![Word::from("Emojis: "),
+///                 Word::from("😂"),
+///                 Word::from("😍")]);
+///
+/// #[cfg(feature = "unicode-linebreak")]
+/// assert_eq!(find_words("CJK: 你好", LineBreakAlgorithm::UnicodeLineBreaks),
+///            vec![Word::from("CJK: "),
+///                 Word::from("你"),
+///                 Word::from("好")]);
+/// ```
+///
+/// A U+2060 Word Joiner character can be inserted if you want to
+/// manually override the defaults and keep the characters together:
+///
+/// ```
+/// use textwrap::core::{find_words, LineBreakAlgorithm, Word};
+///
+/// #[cfg(feature = "unicode-linebreak")]
+/// assert_eq!(find_words("Emojis: 😂\u{2060}😍", LineBreakAlgorithm::UnicodeLineBreaks),
+///            vec![Word::from("Emojis: "),
+///                 Word::from("😂\u{2060}😍")]);
+/// ```
+///
+/// The Unicode line breaking algorithm will suppress break breaks
+/// around certain punctuation characters::
+///
+/// ```
+/// use textwrap::core::{find_words, LineBreakAlgorithm, Word};
+///
+/// #[cfg(feature = "unicode-linebreak")]
+/// assert_eq!(find_words("[ foo ] bar !", LineBreakAlgorithm::UnicodeLineBreaks),
+///            vec![Word::from("[ foo ] "),
+///                 Word::from("bar !")]);
+/// ```
+pub fn find_words(line: &str, line_break_algorithm: LineBreakAlgorithm) -> Vec<Word> {
+    match line_break_algorithm {
+        LineBreakAlgorithm::Whitespace => find_ascii_words(line).collect::<Vec<_>>(),
+        #[cfg(feature = "unicode-linebreak")]
+        LineBreakAlgorithm::UnicodeLineBreaks => find_unicode_words(line).collect::<Vec<_>>(),
+    }
```

Comment:
This needs to go — I'll introduce a trait instead so we can avoid branches at runtime and instead push them to compile time.

Reply:
I just wanted to comment on this :) Yep, I do also think that a trait is a better idea than an enum here. It just introduces something that the maintainers have to keep track of and change if someone wants their algorithm to get included/be accessible to `textwrap`.

Reply:
Yes, I agree. Mostly because of how a trait will streamline the code. The openness of a trait is a secondary benefit: so far, I don't think anyone has made their own implementation of the `WordSplitter` trait.

Reply:
I've updated the PR with a new `LineBreakAlgorithm` trait. I'm not super happy with the names here, so I might change my mind in the next few days...

Reply:
I changed my mind and we now have a `WordSeparator` trait. I've also reworked the existing `WordSplitter` trait so that that the two traits work in a consistent way, see #331 and #332.